File: Mbs.Api.Host.Ng/ClientApp/src/app/shared/mbs/data/generators/quote-parameters.ts
import { objectName, spreadFractionName, askSizeName, bidSizeName } from './constants';

/** The quote related input parameters for the waveform generators.
 *
 *  ➊ The ask and bid prices are equidistant from the mid price.
 *  (a) The half-length of the spread is defined as a ratio ∈[0, 1].
 *  (b) When the ratio is 1, the bid price is zero and the ask price is twice the mid price.
 *  (c) When the ratio is 0, the both ask and bid prices are equal to the the mid price.
 *
 *  ➋ The ask and bid sizes are constant values. */
export class QuoteParameters {
    private static readonly defaultSpreadFraction: number = 0.3;
    private static readonly defaultAskSize: number = 100;
    private static readonly defaultBidSize: number = 100;

    /** The spread fraction, ρs, which determines the ask and bid prices as a fraction of the mid price. */
    spreadFraction: number = QuoteParameters.defaultSpreadFraction;

    /** The ask size, which is the same for all quotes; should be positive. */
    askSize: number = QuoteParameters.defaultAskSize;

    /** The bid size, which is the same for all quotes; should be positive. */
    bidSize: number = QuoteParameters.defaultBidSize;

    constructor(data?: QuoteParameters) {
        if (data) {
            for (const property in data) {
                if (data.hasOwnProperty(property)) {
                    (<any>this)[property] = (<any>data)[property];
                }
            }
        }
    }

    static fromJS(data: any): QuoteParameters {
        data = typeof data === objectName ? data : {};
        const result = new QuoteParameters();
        result.init(data);
        return result;
    }

    private init(data?: any): void {
        if (data) {
            this.spreadFraction = data[spreadFractionName] !== undefined
                ? data[spreadFractionName] : QuoteParameters.defaultSpreadFraction;
            this.askSize = data[askSizeName] !== undefined
                ? data[askSizeName] : QuoteParameters.defaultAskSize;
            this.bidSize = data[bidSizeName] !== undefined
                ? data[bidSizeName] : QuoteParameters.defaultBidSize;
        }
    }

    toJSON(data?: any) {
        data = typeof data === objectName ? data : {};
        data[spreadFractionName] = this.spreadFraction;
        data[askSizeName] = this.askSize;
        data[bidSizeName] = this.bidSize;
        return data;
    }
}

File: Mbs/Trading/Data/Generators/QuoteParameters.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using Mbs.Trading.Data.Entities;

namespace Mbs.Trading.Data.Generators
{
    /// <summary>
    /// The <see cref="Quote"/> related input parameters for the waveform generators.
    /// A waveform generator produces quotes with the following traits.
    /// <para/>➊ The ask and bid prices are equidistant from the mid price.
    /// <para/>The half-length of the spread is defined as a ratio ∈[0, 1].
    /// <para/>When the ratio is 1, the bid price is zero and the ask price is twice the mid price.
    /// <para/>When the ratio is 0, the both ask and bid prices are equal to the the mid price.
    /// <para/>➋ The ask and bid sizes are constant values.
    /// </summary>
    public class QuoteParameters : IValidatableObject
    {
        /// <summary>
        /// Gets or sets the spread fraction, ρs, which determines the ask and bid prices as a fraction of the mid price.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.SpreadFraction)]
        [Range(0.0, 1.0, ErrorMessage = ErrorMessages.FieldMustBeInRange01)]
        public double SpreadFraction { get; set; } = DefaultParameterValues.SpreadFraction;

        /// <summary>
        /// Gets or sets the value of the ask size, which is the same for all quotes; should be positive.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.AskSize)]
        [Range(0.0, double.MaxValue, ErrorMessage = ErrorMessages.FieldMustBePositive)]
        public double AskSize { get; set; } = DefaultParameterValues.AskSize;

        /// <summary>
        /// Gets or sets the value of the bid size, which is the same for all quotes; should be positive.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.BidSize)]
        [Range(0.0, double.MaxValue, ErrorMessage = ErrorMessages.FieldMustBePositive)]
        public double BidSize { get; set; } = DefaultParameterValues.BidSize;

        /// <inheritdoc />
        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
        {
            yield break;
        }
    }
}
